server/service/queries.go

```diff
@@ -479,6 +477,21 @@ func (svc *Service) ApplyQuerySpecs(ctx context.Context, specs []*fleet.QuerySpe
 				message: fmt.Sprintf("query payload verification: %s", err),
 			})
 		}
+
+		// check that the user can update the query if it already exists
+		query, err := svc.ds.QueryByName(ctx, query.Name)
```

Comment:
The thing here is that we have to load the query to check if it exists (which means it will be an update). The first level of authorization is done above (to check if the user has any authorization to create a new query).

Of course this approach is racy - one of the query names could not exist when we look it up, and exist once the call gets to the datastore layer and does an `insert ... on dupicate update`, but given that authorization should be done at the service layer I think that's good enough for now. A more complex solution would be to split here - at the service level - which ones exist and which ones don't, and turn them into INSERTs and UPDATEs, and at the datastore layer only run that statement (i.e. not an UPSERT), and fail if the INSERT results in a conflict, or an UPDATE does not exist anymore. That way, the authorization would be guaranteed to be correct.

That being said, despite the race risk, this approach is still an improvement on how it currently works.

Reply:
Agreed, I think there is still room for improvement here. 

> Should we authorize first then verify?

In order to perform the authorization check, the name is required so that we can look up the query with the given name. That is why I first performed validation on the query before performing the authorization check.

> Of course this approach is racy - one of the query names could not exist when we look it up, and exist once the call gets to the datastore layer

Additionally, if someone else updates the query at the same time (which updates the author ID on the query), then it's possible that we still allow the update even though the user should no longer be authorized to do so. To solve this we could do a `select ... for update` inside a transaction.

Reply:
> Additionally, if someone else updates the query at the same time (which updates the author ID on the query), then it's possible that we still allow the update even though the user should no longer be authorized to do so.

Yeah, that's true. This one would still be possible if we split in INSERTs/UPDATEs, although of course the datastore layer could still make that check when updating, or as you mention use db-level locking in a transaction. In any case, IMO that would be a follow-up PR in the sense that the current one improves things and keeps it simple, and we can iterate on that and discuss the best approach to improve this later on.

Reply:
Although we _are_ in feature freeze, so maybe we can discuss that at tomorrow's backend meeting before it gets merged!